iree/compiler/Conversion/Common/PadLinalgOps.cpp

```diff
@@ -0,0 +1,151 @@
+    int newKSize = std::ceil(float(K) / paddingSize) * paddingSize;
+
+    int paddingForM = newMSize - M;
+    int paddingForN = newNSize - N;
+    int paddingForK = newKSize - K;
+
+    if (paddingForM == 0 && paddingForN == 0 && paddingForK == 0)
+      return failure();
+
+    auto getPaddedOperand = [&](Value operand, ArrayRef<int64_t> shape,
+                                ArrayRef<int64_t> highPadding) -> Value {
+      if (llvm::all_of(highPadding,
+                       [](int64_t val) -> bool { return val == 0; })) {
+        return operand;
+      }
+      auto elementType =
+          operand.getType().cast<RankedTensorType>().getElementType();
+      auto paddedType = RankedTensorType::get(shape, elementType);
+      Value paddingValue =
+          rewriter.create<ConstantOp>(loc, rewriter.getZeroAttr(elementType));
+
+      auto padTensorOp = rewriter.create<linalg::PadTensorOp>(
+          loc, paddedType, operand, ArrayRef<Value>{}, ArrayRef<Value>{},
+          rewriter.getI64ArrayAttr({0, 0}),
+          rewriter.getI64ArrayAttr(highPadding));
+
+      int rank = padTensorOp.getResultType().getRank();
+      SmallVector<Type, 4> blockArgTypes;
+      blockArgTypes.assign(rank, rewriter.getIndexType());
+      auto &region = padTensorOp.region();
+      OpBuilder::InsertionGuard guard(rewriter);
+      rewriter.createBlock(&region, region.end(), blockArgTypes);
+      rewriter.create<linalg::YieldOp>(loc, paddingValue);
+      return padTensorOp;
+    };
+
+    auto paddedLhs =
+        getPaddedOperand(lhs, {newMSize, newKSize}, {paddingForM, paddingForK});
+
+    auto paddedrhs =
+        getPaddedOperand(rhs, {newKSize, newNSize}, {paddingForK, paddingForN});
+
+    auto resultType = RankedTensorType::get(
+        {newMSize, newNSize},
+        result.getType().cast<RankedTensorType>().getElementType());
+
+    // Padding for K-dim only result doesn't change result size.
+    if (paddingForM == 0 && paddingForN == 0) {
+      auto paddedMatmulOp =
+          cast<linalg::LinalgOp>(matmulOp.getOperation())
+              .clone(rewriter, loc, {resultType},
+                     ArrayRef<Value>{paddedLhs, paddedrhs, result});
+      rewriter.replaceOp(matmulOp, paddedMatmulOp->getResults());
+    } else {
+      auto paddedResult = getPaddedOperand(result, {newMSize, newNSize},
+                                           {paddingForM, paddingForN});
+      auto paddedMatmulOp =
+          cast<linalg::LinalgOp>(matmulOp.getOperation())
+              .clone(rewriter, loc, {resultType},
+                     ArrayRef<Value>{paddedLhs, paddedrhs, paddedResult});
+
+      SmallVector<OpFoldResult> offsets(2, rewriter.getI64IntegerAttr(0));
+      SmallVector<OpFoldResult> strides(2, rewriter.getI64IntegerAttr(1));
+      SmallVector<OpFoldResult> sizes = {rewriter.getIndexAttr(M),
+                                         rewriter.getIndexAttr(N)};
+      rewriter.replaceOpWithNewOp<SubTensorOp>(
```

Comment:
Would be good to see the performance impact of this `SubTensorOp`. There shouldnt be any as it should be fused with its consumers.

Reply:
It doesn't to have significant implications (at least at when measured the MobileNetV2/V3 benchmarks on CPU comparing to padding workgroup tiles)